// @primer/react/dist/ActionBar/ActionBar.js
import { c } from 'react-compiler-runtime';
import React, { useRef, useState, forwardRef, useId } from 'react';
import { KebabHorizontalIcon } from '@primer/octicons-react';
import { ActionList } from '../ActionList/index.js';
import useIsomorphicLayoutEffect from '../utils/useIsomorphicLayoutEffect.js';
import { useResizeObserver } from '../hooks/useResizeObserver.js';
import { IconButton } from '../Button/IconButton.js';
import { useFocusZone } from '../hooks/useFocusZone.js';
import styles from './ActionBar.module.css.js';
import { clsx } from 'clsx';
import { jsxs, jsx } from 'react/jsx-runtime';
import { useRefObjectAsForwardedRef } from '../hooks/useRefObjectAsForwardedRef.js';
import { FocusKeys } from '@primer/behaviors';
import { ActionMenu } from '../ActionMenu/ActionMenu.js';

const ACTIONBAR_ITEM_GAP = 8;

/**
 * Registry of descendants to render in the list or menu. To preserve insertion order across updates, children are
 * set to `null` when unregistered rather than fully dropped from the map.
 */

const ActionBarContext = /*#__PURE__*/React.createContext({
  size: 'medium',
  registerChild: () => {},
  unregisterChild: () => {},
  isVisibleChild: () => true,
  groupId: undefined
});

/*
small (28px), medium (32px), large (40px)
*/

const MORE_BTN_WIDTH = 32;
const calculatePossibleItems = (registryEntries, navWidth, gap, moreMenuWidth = 0) => {
  const widthToFit = navWidth - moreMenuWidth;
  let breakpoint = registryEntries.length; // assume all items will fit
  let sumsOfChildWidth = 0;
  for (const [index, [, child]] of registryEntries.entries()) {
    sumsOfChildWidth += index > 0 ? child.width + gap : child.width;
    if (sumsOfChildWidth > widthToFit) {
      breakpoint = index;
      break;
    } else {
      continue;
    }
  }
  return breakpoint;
};
const renderMenuItem = (item, index) => {
  if (item.type === 'divider') {
    return /*#__PURE__*/jsx(ActionList.Divider, {}, index);
  }
  const {
    label,
    onClick,
    disabled,
    trailingVisual: TrailingIcon,
    leadingVisual: LeadingIcon,
    items,
    variant
  } = item;
  if (items && items.length > 0) {
    return /*#__PURE__*/jsxs(ActionMenu, {
      children: [/*#__PURE__*/jsx(ActionMenu.Anchor, {
        children: /*#__PURE__*/jsxs(ActionList.Item, {
          disabled: disabled,
          variant: variant,
          children: [LeadingIcon ? /*#__PURE__*/jsx(ActionList.LeadingVisual, {
            children: /*#__PURE__*/jsx(LeadingIcon, {})
          }) : null, label, TrailingIcon ? /*#__PURE__*/jsx(ActionList.TrailingVisual, {
            children: typeof TrailingIcon === 'string' ? /*#__PURE__*/jsx("span", {
              children: TrailingIcon
            }) : /*#__PURE__*/jsx(TrailingIcon, {})
          }) : null]
        })
      }), /*#__PURE__*/jsx(ActionMenu.Overlay, {
        children: /*#__PURE__*/jsx(ActionList, {
          children: items.map((subItem, subIndex) => renderMenuItem(subItem, subIndex))
        })
      })]
    }, label);
  }
  return /*#__PURE__*/jsxs(ActionList.Item, {
    onSelect: onClick,
    disabled: disabled,
    variant: variant,
    children: [LeadingIcon ? /*#__PURE__*/jsx(ActionList.LeadingVisual, {
      children: /*#__PURE__*/jsx(LeadingIcon, {})
    }) : null, label, TrailingIcon ? /*#__PURE__*/jsx(ActionList.TrailingVisual, {
      children: typeof TrailingIcon === 'string' ? /*#__PURE__*/jsx("span", {
        children: TrailingIcon
      }) : /*#__PURE__*/jsx(TrailingIcon, {})
    }) : null]
  }, label);
};
renderMenuItem.displayName = "renderMenuItem";
const getMenuItems = (navWidth, moreMenuWidth, childRegistry, hasActiveMenu, gap) => {
  const registryEntries = Array.from(childRegistry).filter(entry => entry[1] !== null && (entry[1].type !== 'action' || entry[1].groupId === undefined));
  if (registryEntries.length === 0) return new Set();
  const numberOfItemsPossible = calculatePossibleItems(registryEntries, navWidth, gap);
  const numberOfItemsPossibleWithMoreMenu = calculatePossibleItems(registryEntries, navWidth, gap, moreMenuWidth || MORE_BTN_WIDTH);
  const menuItems = new Set();

  // First, we check if we can fit all the items with their icons
  if (registryEntries.length >= numberOfItemsPossible) {
    /* Below is an accessibility requirement. Never show only one item in the overflow menu.
     * If there is only one item left to display in the overflow menu according to the calculation,
     * we need to pull another item from the list into the overflow menu.
     */
    const numberOfItemsInMenu = registryEntries.length - numberOfItemsPossibleWithMoreMenu;
    const numberOfListItems = numberOfItemsInMenu === 1 ? numberOfItemsPossibleWithMoreMenu - 1 : numberOfItemsPossibleWithMoreMenu;
    for (const [index, [id, child]] of registryEntries.entries()) {
      if (index < numberOfListItems) {
        continue;
        //if the last item is a divider
      } else if (child.type === 'divider') {
        if (index === numberOfListItems - 1 || index === numberOfListItems) {
          continue;
        } else {
          menuItems.add(id);
        }
      } else {
        menuItems.add(id);
      }
    }
    return menuItems;
  } else if (numberOfItemsPossible > registryEntries.length && hasActiveMenu) {
    /* If the items fit in the list and there are items in the overflow menu, we need to move them back to the list */
    return new Set();
  }
};
const ActionBar = props => {
  const $ = c(37);
  const {
    size: t0,
    children,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    flush: t1,
    className,
    gap: t2
  } = props;
  const size = t0 === undefined ? "medium" : t0;
  const flush = t1 === undefined ? false : t1;
  const gap = t2 === undefined ? "condensed" : t2;
  const listRef = useRef(null);
  const [computedGap, setComputedGap] = useState(ACTIONBAR_ITEM_GAP);
  const [childRegistry, setChildRegistry] = useState(_temp);
  let t3;
  if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
    t3 = (id, childProps) => setChildRegistry(prev => new Map(prev).set(id, childProps));
    $[0] = t3;
  } else {
    t3 = $[0];
  }
  const registerChild = t3;
  let t4;
  if ($[1] === Symbol.for("react.memo_cache_sentinel")) {
    t4 = id_0 => setChildRegistry(prev_0 => new Map(prev_0).set(id_0, null));
    $[1] = t4;
  } else {
    t4 = $[1];
  }
  const unregisterChild = t4;
  const [menuItemIds, setMenuItemIds] = useState(_temp2);
  const navRef = useRef(null);
  let t5;
  if ($[2] === Symbol.for("react.memo_cache_sentinel")) {
    t5 = () => {
      if (!listRef.current) {
        return;
      }
      const g = window.getComputedStyle(listRef.current).gap;
      const parsed = parseFloat(g);
      if (!Number.isNaN(parsed)) {
        setComputedGap(parsed);
      }
    };
    $[2] = t5;
  } else {
    t5 = $[2];
  }
  let t6;
  if ($[3] !== gap) {
    t6 = [gap];
    $[3] = gap;
    $[4] = t6;
  } else {
    t6 = $[4];
  }
  useIsomorphicLayoutEffect(t5, t6);
  const moreMenuRef = useRef(null);
  let t7;
  if ($[5] !== childRegistry || $[6] !== computedGap || $[7] !== menuItemIds.size) {
    t7 = resizeObserverEntries => {
      var _moreMenuRef$current$, _moreMenuRef$current;
      const navWidth = resizeObserverEntries[0].contentRect.width;
      const moreMenuWidth = (_moreMenuRef$current$ = (_moreMenuRef$current = moreMenuRef.current) === null || _moreMenuRef$current === void 0 ? void 0 : _moreMenuRef$current.getBoundingClientRect().width) !== null && _moreMenuRef$current$ !== void 0 ? _moreMenuRef$current$ : 0;
      const hasActiveMenu = menuItemIds.size > 0;
      if (navWidth > 0) {
        const newMenuItemIds = getMenuItems(navWidth, moreMenuWidth, childRegistry, hasActiveMenu, computedGap);
        if (newMenuItemIds) {
          setMenuItemIds(newMenuItemIds);
        }
      }
    };
    $[5] = childRegistry;
    $[6] = computedGap;
    $[7] = menuItemIds.size;
    $[8] = t7;
  } else {
    t7 = $[8];
  }
  useResizeObserver(t7, navRef);
  let t8;
  if ($[9] !== menuItemIds) {
    t8 = id_1 => !menuItemIds.has(id_1);
    $[9] = menuItemIds;
    $[10] = t8;
  } else {
    t8 = $[10];
  }
  const isVisibleChild = t8;
  let t9;
  if ($[11] === Symbol.for("react.memo_cache_sentinel")) {
    t9 = {
      containerRef: listRef,
      bindKeys: FocusKeys.ArrowHorizontal | FocusKeys.HomeAndEnd,
      focusOutBehavior: "wrap"
    };
    $[11] = t9;
  } else {
    t9 = $[11];
  }
  useFocusZone(t9);
  let groupedItemsMap;
  if ($[12] !== childRegistry) {
    groupedItemsMap = new Map();
    for (const [key, childProps_0] of childRegistry) {
      if ((childProps_0 === null || childProps_0 === void 0 ? void 0 : childProps_0.type) === "action" && childProps_0.groupId) {
        const existingGroup = groupedItemsMap.get(childProps_0.groupId) || [];
        existingGroup.push([key, childProps_0]);
        groupedItemsMap.set(childProps_0.groupId, existingGroup);
      }
    }
    $[12] = childRegistry;
    $[13] = groupedItemsMap;
  } else {
    groupedItemsMap = $[13];
  }
  const groupedItems = groupedItemsMap;
  let t10;
  if ($[14] !== isVisibleChild || $[15] !== size) {
    t10 = {
      size,
      registerChild,
      unregisterChild,
      isVisibleChild
    };
    $[14] = isVisibleChild;
    $[15] = size;
    $[16] = t10;
  } else {
    t10 = $[16];
  }
  let t11;
  if ($[17] !== className) {
    t11 = clsx(className, styles.Nav);
    $[17] = className;
    $[18] = t11;
  } else {
    t11 = $[18];
  }
  let t12;
  if ($[19] !== ariaLabel || $[20] !== childRegistry || $[21] !== groupedItems || $[22] !== menuItemIds) {
    t12 = menuItemIds.size > 0 && /*#__PURE__*/jsxs(ActionMenu, {
      children: [/*#__PURE__*/jsx(ActionMenu.Anchor, {
        children: /*#__PURE__*/jsx(IconButton, {
          variant: "invisible",
          "aria-label": `More ${ariaLabel} items`,
          icon: KebabHorizontalIcon
        })
      }), /*#__PURE__*/jsx(ActionMenu.Overlay, {
        children: /*#__PURE__*/jsx(ActionList, {
          children: Array.from(menuItemIds).map(id_2 => {
            const menuItem = childRegistry.get(id_2);
            if (!menuItem) {
              return null;
            }
            if (menuItem.type === "divider") {
              return /*#__PURE__*/jsx(ActionList.Divider, {}, id_2);
            } else {
              if (menuItem.type === "action") {
                const {
                  onClick,
                  icon: Icon,
                  label,
                  disabled
                } = menuItem;
                return /*#__PURE__*/jsxs(ActionList.Item, {
                  onSelect: event => {
                    typeof onClick === "function" && onClick(event);
                  },
                  disabled: disabled,
                  children: [/*#__PURE__*/jsx(ActionList.LeadingVisual, {
                    children: /*#__PURE__*/jsx(Icon, {})
                  }), label]
                }, label);
              }
            }
            if (menuItem.type === "menu") {
              const menuItems = menuItem.items;
              const {
                icon: Icon_0,
                label: label_0,
                returnFocusRef
              } = menuItem;
              return /*#__PURE__*/jsxs(ActionMenu, {
                children: [/*#__PURE__*/jsx(ActionMenu.Anchor, {
                  children: /*#__PURE__*/jsxs(ActionList.Item, {
                    children: [Icon_0 !== "none" ? /*#__PURE__*/jsx(ActionList.LeadingVisual, {
                      children: /*#__PURE__*/jsx(Icon_0, {})
                    }) : null, label_0]
                  })
                }), /*#__PURE__*/jsx(ActionMenu.Overlay, {
                  ...(returnFocusRef && {
                    returnFocusRef
                  }),
                  children: /*#__PURE__*/jsx(ActionList, {
                    children: menuItems.map(_temp3)
                  })
                })]
              }, id_2);
            }
            const groupedMenuItems = groupedItems.get(id_2) || [];
            if (menuItem.type === "group") {
              return /*#__PURE__*/jsx(React.Fragment, {
                children: groupedMenuItems.map(_temp4)
              }, id_2);
            }
          })
        })
      })]
    });
    $[19] = ariaLabel;
    $[20] = childRegistry;
    $[21] = groupedItems;
    $[22] = menuItemIds;
    $[23] = t12;
  } else {
    t12 = $[23];
  }
  let t13;
  if ($[24] !== ariaLabel || $[25] !== ariaLabelledBy || $[26] !== children || $[27] !== gap || $[28] !== t12) {
    t13 = /*#__PURE__*/jsxs("div", {
      ref: listRef,
      role: "toolbar",
      className: styles.List,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledBy,
      "data-gap": gap,
      children: [children, t12]
    });
    $[24] = ariaLabel;
    $[25] = ariaLabelledBy;
    $[26] = children;
    $[27] = gap;
    $[28] = t12;
    $[29] = t13;
  } else {
    t13 = $[29];
  }
  let t14;
  if ($[30] !== flush || $[31] !== t11 || $[32] !== t13) {
    t14 = /*#__PURE__*/jsx("div", {
      ref: navRef,
      className: t11,
      "data-flush": flush,
      children: t13
    });
    $[30] = flush;
    $[31] = t11;
    $[32] = t13;
    $[33] = t14;
  } else {
    t14 = $[33];
  }
  let t15;
  if ($[34] !== t10 || $[35] !== t14) {
    t15 = /*#__PURE__*/jsx(ActionBarContext.Provider, {
      value: t10,
      children: t14
    });
    $[34] = t10;
    $[35] = t14;
    $[36] = t15;
  } else {
    t15 = $[36];
  }
  return t15;
};
const ActionBarIconButton = /*#__PURE__*/forwardRef((t0, forwardedRef) => {
  const $ = c(27);
  let disabled;
  let onClick;
  let props;
  if ($[0] !== t0) {
    ({
      disabled,
      onClick,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = disabled;
    $[2] = onClick;
    $[3] = props;
  } else {
    disabled = $[1];
    onClick = $[2];
    props = $[3];
  }
  const ref = useRef(null);
  useRefObjectAsForwardedRef(forwardedRef, ref);
  const id = useId();
  const {
    size,
    registerChild,
    unregisterChild,
    isVisibleChild
  } = React.useContext(ActionBarContext);
  const {
    groupId
  } = React.useContext(ActionBarGroupContext);
  const widthRef = useRef();
  let t1;
  if ($[4] !== disabled || $[5] !== groupId || $[6] !== id || $[7] !== onClick || $[8] !== props || $[9] !== registerChild || $[10] !== unregisterChild) {
    t1 = () => {
      var _ref$current, _props$ariaLabel;
      const width = (_ref$current = ref.current) === null || _ref$current === void 0 ? void 0 : _ref$current.getBoundingClientRect().width;
      if (width) {
        widthRef.current = width;
      }
      if (!widthRef.current) {
        return;
      }
      registerChild(id, {
        type: "action",
        label: (_props$ariaLabel = props["aria-label"]) !== null && _props$ariaLabel !== void 0 ? _props$ariaLabel : "",
        icon: props.icon,
        disabled: !!disabled,
        onClick: onClick,
        width: widthRef.current,
        groupId: groupId !== null && groupId !== void 0 ? groupId : undefined
      });
      return () => {
        unregisterChild(id);
      };
    };
    $[4] = disabled;
    $[5] = groupId;
    $[6] = id;
    $[7] = onClick;
    $[8] = props;
    $[9] = registerChild;
    $[10] = unregisterChild;
    $[11] = t1;
  } else {
    t1 = $[11];
  }
  const t2 = props["aria-label"];
  let t3;
  if ($[12] !== disabled || $[13] !== onClick || $[14] !== props.icon || $[15] !== registerChild || $[16] !== t2 || $[17] !== unregisterChild) {
    t3 = [registerChild, unregisterChild, t2, props.icon, disabled, onClick];
    $[12] = disabled;
    $[13] = onClick;
    $[14] = props.icon;
    $[15] = registerChild;
    $[16] = t2;
    $[17] = unregisterChild;
    $[18] = t3;
  } else {
    t3 = $[18];
  }
  useIsomorphicLayoutEffect(t1, t3);
  let t4;
  if ($[19] !== disabled || $[20] !== onClick) {
    t4 = event => {
      var _onClick;
      if (disabled) {
        return;
      }
      (_onClick = onClick) === null || _onClick === void 0 ? void 0 : _onClick(event);
    };
    $[19] = disabled;
    $[20] = onClick;
    $[21] = t4;
  } else {
    t4 = $[21];
  }
  const clickHandler = t4;
  if (!isVisibleChild(id) || groupId && !isVisibleChild(groupId)) {
    return null;
  }
  let t5;
  if ($[22] !== clickHandler || $[23] !== disabled || $[24] !== props || $[25] !== size) {
    t5 = /*#__PURE__*/jsx(IconButton, {
      "aria-disabled": disabled,
      ref: ref,
      size: size,
      onClick: clickHandler,
      ...props,
      variant: "invisible"
    });
    $[22] = clickHandler;
    $[23] = disabled;
    $[24] = props;
    $[25] = size;
    $[26] = t5;
  } else {
    t5 = $[26];
  }
  return t5;
});
const ActionBarGroupContext = /*#__PURE__*/React.createContext({
  groupId: null
});
const ActionBarGroup = /*#__PURE__*/forwardRef((t0, forwardedRef) => {
  const $ = c(16);
  const {
    children
  } = t0;
  const backupRef = useRef(null);
  const ref = forwardedRef !== null && forwardedRef !== void 0 ? forwardedRef : backupRef;
  const id = useId();
  const {
    registerChild,
    unregisterChild
  } = React.useContext(ActionBarContext);
  const widthRef = useRef();
  let t1;
  if ($[0] !== id || $[1] !== ref || $[2] !== registerChild || $[3] !== unregisterChild) {
    t1 = () => {
      var _ref$current2;
      const width = (_ref$current2 = ref.current) === null || _ref$current2 === void 0 ? void 0 : _ref$current2.getBoundingClientRect().width;
      if (width) {
        widthRef.current = width;
      }
      if (!widthRef.current) {
        return;
      }
      registerChild(id, {
        type: "group",
        width: widthRef.current
      });
      return () => {
        unregisterChild(id);
      };
    };
    $[0] = id;
    $[1] = ref;
    $[2] = registerChild;
    $[3] = unregisterChild;
    $[4] = t1;
  } else {
    t1 = $[4];
  }
  let t2;
  if ($[5] !== registerChild || $[6] !== unregisterChild) {
    t2 = [registerChild, unregisterChild];
    $[5] = registerChild;
    $[6] = unregisterChild;
    $[7] = t2;
  } else {
    t2 = $[7];
  }
  useIsomorphicLayoutEffect(t1, t2);
  let t3;
  if ($[8] !== id) {
    t3 = {
      groupId: id
    };
    $[8] = id;
    $[9] = t3;
  } else {
    t3 = $[9];
  }
  let t4;
  if ($[10] !== children || $[11] !== ref) {
    t4 = /*#__PURE__*/jsx("div", {
      className: styles.Group,
      ref: ref,
      children: children
    });
    $[10] = children;
    $[11] = ref;
    $[12] = t4;
  } else {
    t4 = $[12];
  }
  let t5;
  if ($[13] !== t3 || $[14] !== t4) {
    t5 = /*#__PURE__*/jsx(ActionBarGroupContext.Provider, {
      value: t3,
      children: t4
    });
    $[13] = t3;
    $[14] = t4;
    $[15] = t5;
  } else {
    t5 = $[15];
  }
  return t5;
});
const ActionBarMenu = /*#__PURE__*/forwardRef((t0, forwardedRef) => {
  const $ = c(42);
  let ariaLabel;
  let icon;
  let items;
  let overflowIcon;
  let props;
  let returnFocusRef;
  if ($[0] !== t0) {
    ({
      "aria-label": ariaLabel,
      icon,
      overflowIcon,
      items,
      returnFocusRef,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = ariaLabel;
    $[2] = icon;
    $[3] = items;
    $[4] = overflowIcon;
    $[5] = props;
    $[6] = returnFocusRef;
  } else {
    ariaLabel = $[1];
    icon = $[2];
    items = $[3];
    overflowIcon = $[4];
    props = $[5];
    returnFocusRef = $[6];
  }
  const backupRef = useRef(null);
  const ref = forwardedRef !== null && forwardedRef !== void 0 ? forwardedRef : backupRef;
  const id = useId();
  const {
    registerChild,
    unregisterChild,
    isVisibleChild
  } = React.useContext(ActionBarContext);
  const [menuOpen, setMenuOpen] = useState(false);
  const widthRef = useRef();
  let t1;
  if ($[7] !== ariaLabel || $[8] !== icon || $[9] !== id || $[10] !== items || $[11] !== overflowIcon || $[12] !== ref || $[13] !== registerChild || $[14] !== returnFocusRef || $[15] !== unregisterChild) {
    t1 = () => {
      var _ref$current3;
      const width = (_ref$current3 = ref.current) === null || _ref$current3 === void 0 ? void 0 : _ref$current3.getBoundingClientRect().width;
      if (width) {
        widthRef.current = width;
      }
      if (!widthRef.current) {
        return;
      }
      registerChild(id, {
        type: "menu",
        width: widthRef.current,
        label: ariaLabel,
        icon: overflowIcon ? overflowIcon : icon,
        returnFocusRef,
        items
      });
      return () => {
        unregisterChild(id);
      };
    };
    $[7] = ariaLabel;
    $[8] = icon;
    $[9] = id;
    $[10] = items;
    $[11] = overflowIcon;
    $[12] = ref;
    $[13] = registerChild;
    $[14] = returnFocusRef;
    $[15] = unregisterChild;
    $[16] = t1;
  } else {
    t1 = $[16];
  }
  let t2;
  if ($[17] !== ariaLabel || $[18] !== icon || $[19] !== items || $[20] !== overflowIcon || $[21] !== registerChild || $[22] !== unregisterChild) {
    t2 = [registerChild, unregisterChild, ariaLabel, overflowIcon, icon, items];
    $[17] = ariaLabel;
    $[18] = icon;
    $[19] = items;
    $[20] = overflowIcon;
    $[21] = registerChild;
    $[22] = unregisterChild;
    $[23] = t2;
  } else {
    t2 = $[23];
  }
  useIsomorphicLayoutEffect(t1, t2);
  if (!isVisibleChild(id)) {
    return null;
  }
  let t3;
  if ($[24] !== ariaLabel || $[25] !== icon || $[26] !== props) {
    t3 = /*#__PURE__*/jsx(ActionMenu.Anchor, {
      children: /*#__PURE__*/jsx(IconButton, {
        variant: "invisible",
        "aria-label": ariaLabel,
        icon: icon,
        ...props
      })
    });
    $[24] = ariaLabel;
    $[25] = icon;
    $[26] = props;
    $[27] = t3;
  } else {
    t3 = $[27];
  }
  let t4;
  if ($[28] !== returnFocusRef) {
    t4 = returnFocusRef && {
      returnFocusRef
    };
    $[28] = returnFocusRef;
    $[29] = t4;
  } else {
    t4 = $[29];
  }
  let t5;
  if ($[30] !== items) {
    t5 = items.map(_temp5);
    $[30] = items;
    $[31] = t5;
  } else {
    t5 = $[31];
  }
  let t6;
  if ($[32] !== t5) {
    t6 = /*#__PURE__*/jsx(ActionList, {
      children: t5
    });
    $[32] = t5;
    $[33] = t6;
  } else {
    t6 = $[33];
  }
  let t7;
  if ($[34] !== t4 || $[35] !== t6) {
    t7 = /*#__PURE__*/jsx(ActionMenu.Overlay, {
      ...t4,
      children: t6
    });
    $[34] = t4;
    $[35] = t6;
    $[36] = t7;
  } else {
    t7 = $[36];
  }
  let t8;
  if ($[37] !== menuOpen || $[38] !== ref || $[39] !== t3 || $[40] !== t7) {
    t8 = /*#__PURE__*/jsxs(ActionMenu, {
      anchorRef: ref,
      open: menuOpen,
      onOpenChange: setMenuOpen,
      children: [t3, t7]
    });
    $[37] = menuOpen;
    $[38] = ref;
    $[39] = t3;
    $[40] = t7;
    $[41] = t8;
  } else {
    t8 = $[41];
  }
  return t8;
});
const VerticalDivider = () => {
  const $ = c(8);
  const ref = useRef(null);
  const id = useId();
  const {
    registerChild,
    unregisterChild,
    isVisibleChild
  } = React.useContext(ActionBarContext);
  const widthRef = useRef();
  let t0;
  if ($[0] !== id || $[1] !== registerChild || $[2] !== unregisterChild) {
    t0 = () => {
      var _ref$current4;
      const width = (_ref$current4 = ref.current) === null || _ref$current4 === void 0 ? void 0 : _ref$current4.getBoundingClientRect().width;
      if (width) {
        widthRef.current = width;
      }
      if (!widthRef.current) {
        return;
      }
      registerChild(id, {
        type: "divider",
        width: widthRef.current
      });
      return () => unregisterChild(id);
    };
    $[0] = id;
    $[1] = registerChild;
    $[2] = unregisterChild;
    $[3] = t0;
  } else {
    t0 = $[3];
  }
  let t1;
  if ($[4] !== registerChild || $[5] !== unregisterChild) {
    t1 = [registerChild, unregisterChild];
    $[4] = registerChild;
    $[5] = unregisterChild;
    $[6] = t1;
  } else {
    t1 = $[6];
  }
  useIsomorphicLayoutEffect(t0, t1);
  if (!isVisibleChild(id)) {
    return null;
  }
  let t2;
  if ($[7] === Symbol.for("react.memo_cache_sentinel")) {
    t2 = /*#__PURE__*/jsx("div", {
      ref: ref,
      "data-component": "ActionBar.VerticalDivider",
      "aria-hidden": "true",
      className: styles.Divider
    });
    $[7] = t2;
  } else {
    t2 = $[7];
  }
  return t2;
};
function _temp() {
  return new Map();
}
function _temp2() {
  return new Set();
}
function _temp3(item, index) {
  return renderMenuItem(item, index);
}
function _temp4(t0) {
  const [key_0, childProps_1] = t0;
  if (childProps_1.type === "action") {
    const {
      onClick: onClick_0,
      icon: Icon_1,
      label: label_1,
      disabled: disabled_0
    } = childProps_1;
    return /*#__PURE__*/jsxs(ActionList.Item, {
      onSelect: event_0 => {
        typeof onClick_0 === "function" && onClick_0(event_0);
      },
      disabled: disabled_0,
      children: [/*#__PURE__*/jsx(ActionList.LeadingVisual, {
        children: /*#__PURE__*/jsx(Icon_1, {})
      }), label_1]
    }, key_0);
  }
  return null;
}
function _temp5(item, index) {
  return renderMenuItem(item, index);
}

export { ActionBar, ActionBarGroup, ActionBarIconButton, ActionBarMenu, VerticalDivider };
