// @primer/react/dist/ActionList/LinkItem.js
import { c } from 'react-compiler-runtime';
import React from 'react';
import { Item } from './Item.js';
import { jsx } from 'react/jsx-runtime';
import Link from '../Link/Link.js';

const LinkItem = /*#__PURE__*/React.forwardRef((t0, forwardedRef) => {
  const $ = c(22);
  let Component;
  let active;
  let className;
  let inactiveText;
  let props;
  let size;
  let variant;
  if ($[0] !== t0) {
    ({
      active,
      inactiveText,
      variant,
      size,
      as: Component,
      className,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = Component;
    $[2] = active;
    $[3] = className;
    $[4] = inactiveText;
    $[5] = props;
    $[6] = size;
    $[7] = variant;
  } else {
    Component = $[1];
    active = $[2];
    className = $[3];
    inactiveText = $[4];
    props = $[5];
    size = $[6];
    variant = $[7];
  }
  const t1 = inactiveText ? true : undefined;
  let t2;
  if ($[8] !== Component || $[9] !== forwardedRef || $[10] !== inactiveText || $[11] !== props) {
    t2 = t3 => {
      const {
        children,
        onClick,
        ...rest
      } = t3;
      const clickHandler = event => {
        onClick && onClick(event);
        props.onClick && props.onClick(event);
      };
      return inactiveText ? /*#__PURE__*/jsx("span", {
        ...rest,
        children: children
      }) : /*#__PURE__*/jsx(Link, {
        as: Component,
        ...rest,
        ...props,
        onClick: clickHandler,
        ref: forwardedRef,
        children: children
      });
    };
    $[8] = Component;
    $[9] = forwardedRef;
    $[10] = inactiveText;
    $[11] = props;
    $[12] = t2;
  } else {
    t2 = $[12];
  }
  let t3;
  if ($[13] !== active || $[14] !== className || $[15] !== inactiveText || $[16] !== props.children || $[17] !== size || $[18] !== t1 || $[19] !== t2 || $[20] !== variant) {
    t3 = /*#__PURE__*/jsx(Item, {
      className: className,
      active: active,
      inactiveText: inactiveText,
      "data-inactive": t1,
      variant: variant,
      size: size,
      _PrivateItemWrapper: t2,
      children: props.children
    });
    $[13] = active;
    $[14] = className;
    $[15] = inactiveText;
    $[16] = props.children;
    $[17] = size;
    $[18] = t1;
    $[19] = t2;
    $[20] = variant;
    $[21] = t3;
  } else {
    t3 = $[21];
  }
  return t3;
});
LinkItem.displayName = 'ActionList.LinkItem';
LinkItem.__SLOT__ = Symbol('ActionList.LinkItem');

export { LinkItem };
