// @primer/react/dist/ActionList/List.js
import { c } from 'react-compiler-runtime';
import React from 'react';
import { fixedForwardRef } from '../utils/modern-polymorphic.js';
import { ActionListContainerContext } from './ActionListContainerContext.js';
import { useSlots } from '../hooks/useSlots.js';
import { Heading } from './Heading.js';
import { useId } from '../hooks/useId.js';
import { ListContext } from './shared.js';
import { useFocusZone } from '../hooks/useFocusZone.js';
import { clsx } from 'clsx';
import classes from './ActionList.module.css.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { useProvidedRefOrCreate } from '../hooks/useProvidedRefOrCreate.js';
import { FocusKeys } from '@primer/behaviors';

const UnwrappedList = (props, forwardedRef) => {
  var _slots$heading$props$;
  const $ = c(37);
  let className;
  let restProps;
  let role;
  let selectionVariant;
  let t0;
  let t1;
  let t2;
  let t3;
  if ($[0] !== props) {
    ({
      as: t0,
      variant: t1,
      selectionVariant,
      showDividers: t2,
      role,
      disableFocusZone: t3,
      className,
      ...restProps
    } = props);
    $[0] = props;
    $[1] = className;
    $[2] = restProps;
    $[3] = role;
    $[4] = selectionVariant;
    $[5] = t0;
    $[6] = t1;
    $[7] = t2;
    $[8] = t3;
  } else {
    className = $[1];
    restProps = $[2];
    role = $[3];
    selectionVariant = $[4];
    t0 = $[5];
    t1 = $[6];
    t2 = $[7];
    t3 = $[8];
  }
  const Component = t0 === undefined ? "ul" : t0;
  const variant = t1 === undefined ? "inset" : t1;
  const showDividers = t2 === undefined ? false : t2;
  const disableFocusZone = t3 === undefined ? false : t3;
  let t4;
  if ($[9] === Symbol.for("react.memo_cache_sentinel")) {
    t4 = {
      heading: Heading
    };
    $[9] = t4;
  } else {
    t4 = $[9];
  }
  const [slots, childrenWithoutSlots] = useSlots(restProps.children, t4);
  const headingId = useId();
  const {
    listRole: listRoleFromContainer,
    listLabelledBy,
    selectionVariant: containerSelectionVariant,
    enableFocusZone: enableFocusZoneFromContainer,
    container
  } = React.useContext(ActionListContainerContext);
  const ariaLabelledBy = slots.heading ? (_slots$heading$props$ = slots.heading.props.id) !== null && _slots$heading$props$ !== void 0 ? _slots$heading$props$ : headingId : listLabelledBy;
  const listRole = role || listRoleFromContainer;
  const listRef = useProvidedRefOrCreate(forwardedRef);
  let enableFocusZone = false;
  if (enableFocusZoneFromContainer !== undefined) {
    enableFocusZone = enableFocusZoneFromContainer;
  } else {
    if (listRole && !disableFocusZone) {
      let t5;
      if ($[10] === Symbol.for("react.memo_cache_sentinel")) {
        t5 = ["menu", "menubar", "listbox"];
        $[10] = t5;
      } else {
        t5 = $[10];
      }
      enableFocusZone = t5.includes(listRole);
    }
  }
  const t5 = !enableFocusZone;
  const t6 = listRole === "menu" || container === "SelectPanel" || container === "FilteredActionList" ? "wrap" : undefined;
  let t7;
  if ($[11] !== listRef || $[12] !== t5 || $[13] !== t6) {
    t7 = {
      disabled: t5,
      containerRef: listRef,
      bindKeys: FocusKeys.ArrowVertical | FocusKeys.HomeAndEnd | FocusKeys.PageUpDown,
      focusOutBehavior: t6
    };
    $[11] = listRef;
    $[12] = t5;
    $[13] = t6;
    $[14] = t7;
  } else {
    t7 = $[14];
  }
  useFocusZone(t7);
  const t8 = selectionVariant || containerSelectionVariant;
  let t9;
  if ($[15] !== headingId || $[16] !== listRole || $[17] !== showDividers || $[18] !== t8 || $[19] !== variant) {
    t9 = {
      variant,
      selectionVariant: t8,
      showDividers,
      role: listRole,
      headingId
    };
    $[15] = headingId;
    $[16] = listRole;
    $[17] = showDividers;
    $[18] = t8;
    $[19] = variant;
    $[20] = t9;
  } else {
    t9 = $[20];
  }
  const t10 = slots.heading;
  let t11;
  if ($[21] !== className) {
    t11 = clsx(classes.ActionList, className);
    $[21] = className;
    $[22] = t11;
  } else {
    t11 = $[22];
  }
  let t12;
  if ($[23] !== Component || $[24] !== ariaLabelledBy || $[25] !== childrenWithoutSlots || $[26] !== listRef || $[27] !== listRole || $[28] !== restProps || $[29] !== showDividers || $[30] !== t11 || $[31] !== variant) {
    t12 = /*#__PURE__*/jsx(Component, {
      className: t11,
      role: listRole,
      "aria-labelledby": ariaLabelledBy,
      ref: listRef,
      "data-dividers": showDividers,
      "data-variant": variant,
      ...restProps,
      children: childrenWithoutSlots
    });
    $[23] = Component;
    $[24] = ariaLabelledBy;
    $[25] = childrenWithoutSlots;
    $[26] = listRef;
    $[27] = listRole;
    $[28] = restProps;
    $[29] = showDividers;
    $[30] = t11;
    $[31] = variant;
    $[32] = t12;
  } else {
    t12 = $[32];
  }
  let t13;
  if ($[33] !== slots.heading || $[34] !== t12 || $[35] !== t9) {
    t13 = /*#__PURE__*/jsxs(ListContext.Provider, {
      value: t9,
      children: [t10, t12]
    });
    $[33] = slots.heading;
    $[34] = t12;
    $[35] = t9;
    $[36] = t13;
  } else {
    t13 = $[36];
  }
  return t13;
};
const List = fixedForwardRef(UnwrappedList);
Object.assign(List, {
  displayName: 'ActionList'
});

export { List };
